### ISSUE TITLE
ntpd-rs doesn't appear to exclude its own address from pools

### ISSUE THREAD

**Issue by paulgear** [pendulum-project/ntpd-rs#1105]
After testing `ntpd-rs` for a few days and being satisfied it's working well, I added it to my internal NTP pool.  A few days afterwards there was a connectivity issue (more on this in #1142) which caused it to re-query DNS and pick up its own address in the response.  Based on the overall and individual figures for dispersion and delay, my suspicion is that it is including itself in its calculations.  Or at the very least, there is no visual indication in `ntp-ctl status` or the metrics that it is excluding its own address.

In the output below, it can be seen that this VM's service address is also in the sources list, and one of the other sources in the pool (2001:44b8:2100:3f00::7b:5) is not being used.  (Count is set to 6, which was the previous maximum number of hosts in the pool.)
```
root@ntp7:~# ntp-ctl status
Synchronization status:
Dispersion: 0.000038s, Delay: 0.001498s
Desired poll interval: 64s
Stratum: 3

Sources:
ntp.lan.gear.systems:123/[2001:44b8:2100:3f00::7b:7]:123 (7): +0.000056±0.000030(±0.000348)s
    pollinterval: 16s, missing polls: 0
    root dispersion: 0.000046s, root delay:0.001495s
ntp.lan.gear.systems:123/[2001:44b8:2100:3f11::7b:6]:123 (8): +0.000035±0.000071(±0.001483)s
    pollinterval: 16s, missing polls: 0
    root dispersion: 0.000031s, root delay:0.000015s
ntp.lan.gear.systems:123/[2001:44b8:2100:3f11::7b:3]:123 (9): -0.000414±0.000094(±0.001989)s
    pollinterval: 16s, missing polls: 0
    root dispersion: 0.001022s, root delay:0.000000s
ntp.lan.gear.systems:123/[2001:44b8:2100:3f11::7b:1]:123 (10): +0.000054±0.000070(±0.001317)s
    pollinterval: 16s, missing polls: 0
    root dispersion: 0.000336s, root delay:0.000549s
ntp.lan.gear.systems:123/[2001:44b8:2100:3f00::7b:4]:123 (11): +0.000218±0.000043(±0.001208)s
    pollinterval: 16s, missing polls: 0
    root dispersion: 0.000458s, root delay:0.000412s
ntp.lan.gear.systems:123/[2001:44b8:2100:3f11::7b:2]:123 (12): +0.000098±0.000076(±0.001398)s
    pollinterval: 16s, missing polls: 0
    root dispersion: 0.000504s, root delay:0.000412s

Servers:
[2001:44b8:2100:3f00::7b:7]:123: received 15583352, accepted 15579996, errors 0
    denied 0, nts nak 0, rate limited 0, ignored 3356
```
Here's a grafana snapshot covering the period in question: https://snapshots.raintank.io/dashboard/snapshot/zQP86qXKrThU011e4fu2yb0wXj01CBoB

To sum up, I think if it's not excluding itself as a source already, it should, and either way it should provide some measurable indication that it has recognised and excluded itself from the list of sources.

**Comment by davidv1992** [comment]
Best I can judge it is excluding itself, the spike in root dispersion is merely the restart effects. We will consider adding explicit display of the exclude-include status in a post 1.0 version.

**Comment by paulgear** [comment]
Thanks David.  I think probably there should also be a mechanism which excludes the local address from counting towards the maximum number of sources for that pool.  If I notice the unreachable behaviour again I'll try to gather some more data.

**Comment by davidv1992** [comment]
Sorry this took a while, implemented a mechanism for this in the form of ignore lists on pools, see #1359.

**Comment by paulgear** [comment]
> Sorry this took a while, implemented a mechanism for this in the form of ignore lists on pools, see #1359.

I don't mean to appear ungrateful, but I don't think #1359 addresses the essence of this issue, which is that any of one's own addresses should always be excluded from use as NTP sources.

Just like NTP resolves loops of order 1 by excluding hosts which are synchronised to itself, it should resolve loops of order 0 by excluding hosts which _are_ itself.  It should do this by enumerating all of the usable addresses on local interfaces and excluding them from consideration as potential sources.

**Comment by davidv1992** [comment]
This is currently not possible to implement well with the current state of loop detection in ntp. Once ntpv5 becomes more widely available this will disappear. Enumerating local interface addresses is not really a solution as this cannot deal with nat arangements and furthermore can cause trouble with more funky configuration setups involving multiple instances of ntp on one machine.

**Comment by davidv1992** [comment]
I get what you mean, but this is given architectural choices we have made quite an effort, and even then tricky particularly when dealing with the generic listen address `0.0.0.0` (or `::`) that we have decided that this is not a priority right now. We still agree with the general advice on pools to not use a pool as source if you are a part of it, and in general consider configurations where it can potentially listen to itself misconfigurations.

If you want to provide a patch providing this we would be open to it however.

**Comment by paulgear** [comment]
Fair enough - there's only so much development effort to go around; for the record, ntpd, chrony, and ntpsec all behave as I would expect in this scenario.

If I were even a novice Rust programmer I would give it a shot myself, but my head is just not in this space enough yet.
